File: course/string/Z.cpp
#include <bits/stdc++.h>

using namespace std;

using ll = int64_t;
using ld = long double;
using ii = pair<int, int>;

const char SPACE = ' ', ENDL = '\n';

//////////////////////////////////////////////////////////////////

template <typename T>
T read() {
    T value;
    cin >> value;
    return value;
}

int ri() {
    return read<int>();
}

ll rll() {
    return read<ll>();
}

string rs() {
    return read<string>();
}

template<typename T>
vector<T> rv(int size) {
    vector<T> a(size);
    for (int i = 0; i < size; ++i) {
        a[i] = read<T>();
    }
    return a;
}

//////////////////////////////////////////////////////////////////

bool yn(bool result, string yes = "YES", string no = "NO") {
	cout << (result ? yes : no) << ENDL;
	return result;
}

//////////////////////////////////////////////////////////////////

vector<string> split(string& s, string seps) {
    vector<string> parts;

    string part = "";
    auto add_part = [&]() {
        parts.push_back(part);
        part = "";
    };

    for (char ch : s) {
        if (string::npos != seps.find(ch)) {
            add_part();
        } else {
            part += ch;
        }
    }

    add_part();

    return parts;
}

//////////////////////////////////////////////////////////////////

ld sqrt_ld(ld value) {
    return sqrt(value);
}

//////////////////////////////////////////////////////////////////

void init_io() {
    string file_name = "";
    if (!file_name.empty()) {
        string input_file_name = file_name + ".in";
        freopen(input_file_name.c_str(), "rt", stdin);

        string output_file_name = file_name + ".out";
        freopen(output_file_name.c_str(), "wt", stdout);
    } else {
        #ifndef ONLINE_JUDGE
        string input_file_name = "../../input.txt"; // reading from file in parent of parent of directory with file
        freopen(input_file_name.c_str(), "rt", stdin);
        #endif
    }

	ios::sync_with_stdio(false);
	cin.tie(nullptr);

	cout << setprecision(10) << fixed;
}

string get_ans(string& s, string& t) {
    vector<int> s_counts(1 << 8, 0);
    for (char ch : s) s_counts[ch]++;

    char none = '\0';
    while (t.length() < s.length()) {
        t += none;
    }

    string no_ans = "";

    string ans(s.length(), none);
    auto can_be_greater = [&](int start_index, char start_ch) {
        ans[start_index] = start_ch;
        --s_counts[start_ch];

        int index = start_index + 1;
        for (char ch = 'z'; ch >= 'a'; --ch) {
            for (int it = 0; it < s_counts[ch]; ++it) {
                ans[index++] = ch;
            }
        }

        bool can = (t < ans);

        ++s_counts[start_ch];
        ans[start_index] = none;

        return can;
    };

    bool prefix_equals = true;
    for (int index = 0; index < s.length(); ++index) {
        char start_ch = 'a';

        char t_ch = t[index];
        if (prefix_equals && none != t_ch) {
            start_ch = t_ch;
        }

        char used_ch = none;

        for (char ch = start_ch; ch <= 'z'; ++ch) {
            if (s_counts[ch] == 0) continue;

            if (can_be_greater(index, ch)) {
                used_ch = ch;
                break;
            }
        }

        if (none == used_ch) return "";

        prefix_equals &= (t_ch == used_ch);
        --s_counts[used_ch];
        ans[index] = used_ch;
    }

    return ans;
}

void solve() {
    string s = rs();
    string t = rs();

    string ans = get_ans(s, t);

    if (ans <= t) ans = "-1";
    cout << ans << ENDL;
}

int main() {
    init_io();
    solve();
    return 0;
}

File: course/string/Z.java
import java.io.BufferedReader;
import java.io.File;
import java.io.FileNotFoundException;
import java.io.FileReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.io.PrintWriter;
import java.util.Arrays;
import java.util.Locale;
import java.util.StringTokenizer;
import java.util.function.BiFunction;

public class Z {

    private void solve() {
        String s = readString();
        String t = readString();

        String answer = getAnswer(s, t);
        if (answer.compareTo(t) <= 0) answer = "-1";
        out.println(answer);
    }

    String getAnswer(String s, String t) {
        char none = '\0';

        if (t.length() < s.length()) {
            t += ("" + none).repeat(s.length() - t.length());
        }

        int[] sCounts = new int[1 << 8];
        for (char ch : s.toCharArray()) sCounts[ch]++;

        char[] answer = new char[s.length()];
        Arrays.fill(answer, none);

        String finalT = t;
        BiFunction<Integer, Character, Boolean> canBeGreater = (startIndex, startCh) -> {
            answer[startIndex] = startCh;
            --sCounts[startCh];

            int index = startIndex + 1;
            for (char ch = 'z'; ch >= 'a'; --ch) {
                for (int it = 0; it < sCounts[ch]; ++it) {
                    answer[index++] = ch;
                }
            }

            boolean can = new String(answer).compareTo(finalT) > 0;

            answer[startIndex] = none;
            ++sCounts[startCh];

            return can;
        };

        boolean prefixEquals = true;
        for (int index = 0; index < answer.length; ++index) {
            char tCh = t.charAt(index);

            char startCh = 'a';
            if (prefixEquals && none != tCh) {
                startCh = tCh;
            }

            char usedCh = none;
            for (char ch = startCh; ch <= 'z'; ++ch) {
                if (0 == sCounts[ch]) continue;

                if (canBeGreater.apply(index, ch)) {
                    usedCh = ch;
                    break;
                }
            }

            if (none == usedCh) {
                return "";
            }

            prefixEquals &= (tCh == usedCh);
            answer[index] = usedCh;
            --sCounts[usedCh];
        }

        return new String(answer);
    }

    //////////////////////////////////////////////////////////////////

    long sqrtLong(long x) {
        long root = (long)Math.sqrt(x);
        while (root * root > x) --root;
        while ((root + 1) * (root + 1) <= x) ++root;
        return root;
    }

    //////////////////////////////////////////////////////////////////

    private boolean yesNo(boolean yes) {
        return yesNo(yes, "YES", "NO");
    }

    private boolean yesNo(boolean yes, String yesString, String noString) {
        out.println(yes ? yesString : noString);
        return yes;
    }

    //////////////////////////////////////////////////////////////////

    private long readLong() {
        return Long.parseLong(readString());
    }

    private int[] readIntArray(int n) {
        int[] a = new int[n];
        for (int i = 0; i < n; ++i) a[i] = readInt();
        return a;
    }

    private int readInt() {
        return Integer.parseInt(readString());
    }

    private String readString() {
        while (!tok.hasMoreTokens()) {
            String nextLine = readLine();
            if (null == nextLine) return null;
            tok = new StringTokenizer(nextLine);
        }

        return tok.nextToken();
    }

    private String readLine() {
        try {
            return in.readLine();
        } catch (IOException e) {
            throw new RuntimeException(e);
        }
    }

    //////////////////////////////////////////////////////////////////

    private BufferedReader in;
    private StringTokenizer tok;
    private PrintWriter out;

    private void initFileIO(String inputFileName, String outputFileName) throws FileNotFoundException {
        in = new BufferedReader(new FileReader(inputFileName));
        out = new PrintWriter(outputFileName);
    }

    private void initConsoleIO() throws IOException {
        in = new BufferedReader(new InputStreamReader(System.in));
        out = new PrintWriter(System.out);
    }

    private void initIO() throws IOException {
        Locale.setDefault(Locale.US);

        String fileName = "";

        if (!fileName.isEmpty()) {
            initFileIO(fileName + ".in", fileName + ".out");
        } else {
            if (new File("input.txt").exists()) {
                initFileIO("input.txt", "output.txt");
            } else {
                initConsoleIO();
            }
        }

        tok = new StringTokenizer("");
    }

    //////////////////////////////////////////////////////////////////

    private void run() {
        try {
            long timeStart = System.currentTimeMillis();

            initIO();
            solve();
            out.close();

            long timeEnd = System.currentTimeMillis();
            System.err.println("Time(ms) = " + (timeEnd - timeStart));
        } catch (Exception e) {
            e.printStackTrace();
            throw new RuntimeException(e);
        }
    }

    public static void main(String[] args) {
        new Z().run();
    }
}
